harness/determined/tensorboard/base.py

```diff
@@ -111,22 +146,22 @@ def __init__(
 
     def run(self) -> None:
         while True:
-            file_paths = self._work_queue.get()
+            task = self._work_queue.get()
 
             # None is the sentinel value
             # to signal the thread to exit
-            if file_paths is None:
+            if task is None:
                 return
 
             # Try-catch is used to avoid exception from
             # one failed sync attempt to cause the thread to exit.
             try:
-                self._upload_function(file_paths)
+                self._upload_function(task.paths, task.mangler, task.rank)
             except Exception as e:
                 logging.warning(f"Sync of Tensorboard files failed with error: {e}")
 
-    def upload(self, paths: List[pathlib.Path]) -> None:
-        self._work_queue.put(paths)
+    def upload(self, task: _UploadTask) -> None:
+        self._work_queue.put(task)
 
     def close(self) -> None:
         self._work_queue.put(None)
```

Comment:
Can you include `self.join()` in close?  Otherwise when `.close()` returns, it doesn't really mean that the object is actually closed.  Also, there's really no case in our system where we'll need to distinguish between "the None has been put on the queue" and "the thread is actually shut down".

Reply:
Actually, I say that, but now that I think about it, it might be nice to do a loop of joins with timeouts, and say something like "we're just waiting for tensorboard to finish uploading" so users don't complain to us about hung trials.